### PR TITLE
neurodamus-models: CMake based neurodamus models

### PR DESCRIPTION
**Pull request by matz-e**

### CODE REVIEW

File: bluebrain/deployment/gitlab-ci.yml [hunk 1]
```diff
@@ -22,6 +22,9 @@ default:
     # We will do this ourselves, will be on GPFS
     GIT_STRATEGY: none
   before_script:
+    # Setup an ominous SSH key to access private Github repos?
+    - eval $(ssh-agent -s)
+    - echo "${BBPHPC_NEURODAMUS_PRIVATE_KEY}" | ssh-add -
```
Comment (pramodk) [inline bluebrain/deployment/gitlab-ci.yml:27]: 🙃 But the temporary thing anyway!

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+from spack.package import *
+
+
+class NeurodamusModels(CMakePackage):
+    """Neuroscientific models to be used with Neurodamus"""
+
+    homepage = "https://github.com/BlueBrain/neurodamus-models"
+    git = "ssh://git@github.com/BlueBrain/neurodamus-models.git"
+
+    version("develop", branch="cfaire")
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]: `just a reminder`: to be changed

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    version("6.6.6", commit="d66d0de8741543497bb127998deba7a5e3f7de42")
+    version("6.6.7", commit="8e21f8e16e037d0f1042d4585b1417f083bd1cea")
+    version("6.6.8", commit="506f1cb7dbf59d9dfca835245597e59e144ed24a")
+    version("6.6.9", commit="2cc59da34d22633df5baff3a2ec282e93de54b97")
+    version("6.7.0", commit="0eee2cc6a4d8f84f3f7980fc6d9c88f5a5f258df")
+    version("6.7.1", commit="3ae2eb94fc8f415d32b89b4003d12b24f4d03384")
+    version("6.7.2", commit="4d89b24a46a36e2b9f1f2d2ee108d7f1c478e64f")
+    version("6.7.3", commit="2dc61b533c1b036c3b52ed0aa14393932d269842")
+    version("6.7.4", commit="b848572f214f799ccb0ccb8615ceb8b925f6dca9")
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]: same

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    variant("caliper", default=False, description="Enable Caliper instrumentation")
+    variant("coreneuron", default=False, description="Enable CoreNEURON support")
+
+    variant(
+        "model",
+        default="neocortex",
+        values=("hippocampus", "neocortex", "thalamus"),
+        multi=False,
+        description="Which mechanisms to enable",
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]: `just nitpick`: 
```suggestion
        description="Which brain region mechanism to enable",
```

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    )
+    variant("ngv", default=False, when="model=neocortex", description="Enable NGV mechanisms")
+    variant(
+        "metabolism", default=False, when="model=neocortex", description="Enable NGV mechanisms"
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]:
```suggestion
        "metabolism", default=False, when="model=neocortex", description="Enable metabolism mechanisms"
```

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    )
+    variant(
+        "plasticity", default=False, when="model=neocortex", description="Enable NGV mechanisms"
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]:
```suggestion
        "plasticity", default=False, when="model=neocortex", description="Enable plasticity related mechanisms"
```

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    )
+    variant("v5", default=True, when="model=neocortex", description="Enable NGV mechanisms")
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]:
```suggestion
    variant("v5", default=True, when="model=neocortex", description="Enable V5 circuit mechanisms")
```

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    def cmake_args(self):
+        args = [
+            self.define("ND_CORE_DIR", self.spec["py-neurodamus"].prefix.lib),
+            self.define_from_variant("ND_MECHANISMS", "model"),
+            self.define_from_variant("ND_ENABLE_CORENEURON", "coreneuron"),
+            self.define_from_variant("ND_NCX_METABOLISM", "metabolism"),
+            self.define_from_variant("ND_NCX_NGV", "ngv"),
+            self.define_from_variant("ND_NCX_PLASTICITY", "plasticity"),
+            self.define_from_variant("ND_NCX_V5", "v5"),
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]: @matz-e: Relatively minor but I am wondering about your opinion here: while reviewing BlueBrain/neurodamus-models/pull/3, I thought having a bit more descriptive name is better than abbreviation. E.g. In this case, `NEURODAMUS_ FOO` is better than `ND_FOO`?

It's a bit verbose but clearer? Once we start using these cmake options in other places like this, later they are cumbersome to change and hence this comment.

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        env.unset("LC_ALL")
+
+    def setup_run_environment(self, env):
+        mech = self.spec.variants["model"].value.upper()
+        env.set(f"NEURODAMUS_{mech}_ROOT", self.prefix)
+        if os.path.isdir(self.prefix.lib.hoc):
+            env.set("HOC_LIBRARY_PATH", self.prefix.lib.hoc)
+        if "+coreneuron" in self.spec:
+            env.set("CORENEURONLIB", self.prefix.lib + "/libcorenrnmech.so")
+        for libnrnmech_name in find(self.prefix.lib, "libnrnmech*.so", recursive=False):
```
Comment (pramodk) [inline bluebrain/repo-bluebrain/packages/neurodamus-models/package.py:71]: these should be platform dependent? i.e. `.dylib` on MacOS?

**Review by pramodk — COMMENTED** [review]
Nice! 👌 

I have few minor comments, otherwise LGTM!

File: bluebrain/deployment/gitlab-ci.yml (hunk 1 above)
Comment (matz-e) [reply bluebrain/deployment/gitlab-ci.yml:27]: Yes. Once the repo is public, this can disappear.

Comment (matz-e) [reply bluebrain/deployment/gitlab-ci.yml:27]: Actually, may want to keep it around for other software in the open-sourcing process

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py (hunk 8 above)
Comment (matz-e) [reply bluebrain/repo-bluebrain/packages/neurodamus-models/package.py]: OK, will do that.

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py (hunk 9 above)
Comment (matz-e) [reply bluebrain/repo-bluebrain/packages/neurodamus-models/package.py:71]: This is more or less ported from the other recipes. Do we ever run the simulation on MacOS?

File: bluebrain/sysconfig/bluebrain5/modules.yaml [hunk 10]
```diff
@@ -13,6 +13,11 @@ modules:
       hash_length: 0
       projections:
         all: '{name}/{version}'
+        neurodamus-models model=hippocampus: 'neurodamus-hippocampus/{version}'
+        neurodamus-models model=neocortex: 'neurodamus-neocortex/{version}'
+        neurodamus-models model=neocortex +plasticity: 'neurodamus-neocortex-plasticity/{version}'
+        neurodamus-models model=neocortex +ngv+metabolism: 'neurodamus-neocortex-multiscale/{version}'
```
Comment (WeinaJi) [inline bluebrain/sysconfig/bluebrain5/modules.yaml]: It seems this PR created modules 
```
neurodamus-models-multiscale/6.7.4
neurodamus-models-plasticity/6.7.4
```
Why not 
```
neurodamus-neocortex-multiscale/6.7.4
neurodamus-neocortex-plasticity/6.7.4
```
?

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py (hunk 9 above)
Comment (pramodk) [reply bluebrain/repo-bluebrain/packages/neurodamus-models/package.py:71]: I believe it was working at least when I was developing stuff (a couple of years ago?) 🙃.
If this is just ported from the current recipe then let's skip for now.

File: bluebrain/sysconfig/bluebrain5/modules.yaml (hunk 10 above)
Comment (matz-e) [reply bluebrain/sysconfig/bluebrain5/modules.yaml]: Thanks for noticing! Indeed, I probably have to fix the projection better.

Comment (matz-e) [reply bluebrain/sysconfig/bluebrain5/modules.yaml]: OK, the `applications_hpc` overwrote the projections I had defined. Should be fixed now.

**Comment by WeinaJi** [comment]
If I remember correctly,  the decision is to tag the models directly from `2.0.0`?  Because the current bb5 models are all 1.x.x. 

Another question, the tag version of py-neurodamus is not shown in the model version. Shall we add or drop it?

**Comment by matz-e** [comment]
> If I remember correctly, the decision is to tag the models directly from `2.0.0`? Because the current bb5 models are all 1.x.x.
> 
> Another question, the tag version of py-neurodamus is not shown in the model version. Shall we add or drop it?

The new 2.0 tag should be produced in ~1 hour. And I would drop the `py-neurodamus` version for simplicity.

File: bluebrain/repo-bluebrain/packages/neurodamus-models/package.py (hunk 9 above)
Comment (ferdonline) [reply bluebrain/repo-bluebrain/packages/neurodamus-models/package.py:71]: @pramodk I remember there was some place that even in Mac neuron used .so extension. (I used to check recipes in mac at some point). So unless you'r sure it's needed I'd refrain from changing

**Comment by WeinaJi** [comment]
> > If I remember correctly, the decision is to tag the models directly from `2.0.0`? Because the current bb5 models are all 1.x.x.
> > Another question, the tag version of py-neurodamus is not shown in the model version. Shall we add or drop it?
> 
> The new 2.0 tag should be produced in ~1 hour. And I would drop the `py-neurodamus` version for simplicity.

That's also good. Thanks!

**Review by WeinaJi — APPROVED** [review]
LGTM! Thanks!

**Review by jorblancoa — APPROVED** [review]

**Pull request merged**